### PR TITLE
[Libmamba] Write debug logs to disk

### PR DESCRIPTION
**Pull request by jaimergp**
During the experimental phase we'll proactively write debug logs to disk to ensure we can asses failures or unexpected behaviours without having to ask for reproducers too much.

The ideas here:

* Write debug logs to `pkgs/cache/.logs/timestamp.log`. This path is generated in the context via a memoized property, so only one path per conda session.
* To support this, we need to rework the configured loggers. Conda relies on the logger objects themselves to set the level, but it's better to configure that through the handler. It also relies on the default handler in the _root_ logger to format everything. Not what we like. Instead, I added the same handler to the `conda` logger. This is fragile so please review with utmost care :) 

Since we haven't touched those files in a while, the autoformatter kicked in BADLY. Disregard the `formatting` commits, sorry!

Tasks:

* [x] Make sure logs are created and populated
* [x] Make sure `-vvv` to stderr still works regardless
* [ ] Make sure everybody likes the banner disclaimer. Investigate other messages too? Maybe at the end of the operation?
* [x] Make sure `libsolv` stderr is also collected in the file.

┆Issue is synchronized with this [Jira Task](https://anaconda.atlassian.net/browse/CONDAORG-25)

### CODE REVIEW

File: tests/common/test_io.py [hunk 1]
```diff
@@ -60,8 +60,8 @@ def test_attach_stderr_handler():
 
     assert len(logr.handlers) == 1
     assert logr.handlers[0].name == 'stderr'
-    assert logr.handlers[0].level is NOTSET
-    assert logr.level is WARN
+    assert logr.handlers[0].level is WARN
```
Comment (jaimergp) [inline tests/common/test_io.py:63]: Discuss thoroughly!

File: tests/common/test_io.py [hunk 2]
```diff
@@ -75,8 +75,8 @@ def test_attach_stderr_handler():
 
     assert len(logr.handlers) == 1
     assert logr.handlers[0].name == 'stderr'
-    assert logr.handlers[0].level is NOTSET
-    assert logr.level is DEBUG
+    assert logr.handlers[0].level is DEBUG
```
Comment (jaimergp) [inline tests/common/test_io.py:78]: Discuss thoroughly!

**Comment by jaimergp** [comment]
Ready for review @jezdez! Note the comments, there are some subtle changes in the logging that need to be discussed team-wide. In general I went for this approach:

Original state:
* `conda` has its own logger that propagates up to the root logger. The logger level is set by context.verbosity.
* The root logger is customized with a stderr handler. The handler level is NOTSET (lowest). Whatever arrives there, will be printed.

This PR:
* `conda` has its own logger but it does not propagate. Its level is NOTSET.
* The custom stderr handler is added to the `conda` logger. Its level is set by context.verbosity.
* Root logger still has the custom stderr handler just in case.
* A new handler has been added to print to a file, with DEBUG level, if libmamba/libmamba2 is in use.
* libsolv stderr is captured and forwarded to the logger.

File: conda/base/context.py [hunk 3]
```diff
@@ -540,6 +540,20 @@ def trash_dir(self):
...
+        # TODO: this inline import can be cleaned up by moving pkgs_dir write detection logic
+        from ..core.package_cache_data import PackageCacheData
+
+        pkgs_dir = PackageCacheData.first_writable().pkgs_dir
+        logs = join(pkgs_dir, ".logs")
+        from ..gateways.disk.create import mkdir_p
+
+        mkdir_p(logs)
+
+        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S-%f")
```
Comment (jezdez) [inline conda/base/context.py:632]: Should this be `datetime.utcnow()`to be able to be comparable?

**Review by jezdez — APPROVED** [review]
Just a few nites, otherwise good to go.

File: conda/base/context.py [hunk 4]
```diff
@@ -540,6 +540,20 @@ def trash_dir(self):
...
         return trash_dir
 
+    @memoizedproperty
+    def logfile_path(self):
+        # TODO: this inline import can be cleaned up by moving pkgs_dir write detection logic
+        from ..core.package_cache_data import PackageCacheData
+
+        pkgs_dir = PackageCacheData.first_writable().pkgs_dir
+        logs = join(pkgs_dir, ".logs")
+        from ..gateways.disk.create import mkdir_p
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda/base/context.py:628]: I didn't check. Only mimicking the function above :/

File: conda/gateways/logging.py [hunk 5]
```diff
@@ -203,23 +205,35 @@ def initialize_std_loggers():
...
     # regardless of the level.
-    attach_stderr_handler(level, 'requests')
-    attach_stderr_handler(level, 'requests.packages.urllib3')
+
+
+@memoize
+def set_file_logging(logger_name=None, level=DEBUG, path=None):
+    if path is None:
+        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
+        os.path.join(f".conda.{timestamp}.log")
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda/gateways/logging.py]: Oops, good catch, that should be `path` and no `join`.

File: conda/gateways/logging.py [hunk 6]
```diff
@@ -203,23 +205,35 @@ def initialize_std_loggers():
...
-    set_conda_log_level(level)  # only set level and use root's handler/formatter
+    set_conda_log_level(level)
     # 'requests' loggers get their own handlers so that they always output messages in long format
     # regardless of the level.
-    attach_stderr_handler(level, 'requests')
-    attach_stderr_handler(level, 'requests.packages.urllib3')
+
+
+@memoize
+def set_file_logging(logger_name=None, level=DEBUG, path=None):
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda/gateways/logging.py:236]: Hm, the [docstring](https://github.com/conda/conda/blob/7a7aa8189342d14dcb1df43ca4a9f9b737c549ce/conda/auxlib/decorators.py#L15-L17) is a bit ambiguous...

>    Decorator to cause a function to cache it's results for each combination of
>    inputs and return the cached result on subsequent calls.  Does not support
>    named arguments or arg values that are not hashable.

Should it read as "named arguments (any kind), or alternatively arg values that are not hashable" or "not-hashable parameters, named or not"? We didn't get any errors though :/

[the remark this answers is not in the document]
Comment (jaimergp) [reply conda/gateways/logging.py:236]: This [line](https://github.com/conda/conda/blob/7a7aa8189342d14dcb1df43ca4a9f9b737c549ce/conda/auxlib/decorators.py#L55) makes me think that named args are ok as long as they are hashable.

**Comment by jaimergp** [comment]
Errors on CI are unrelated to this PR. See #11161.

I addressed your comments @jezdez - let me know what you think! Thanks!

**Review by jezdez — APPROVED** [review]
Nicely done!

**Pull request merged**